doc/dev/mds_internals/quiesce.rst

```diff
@@ -0,0 +1,113 @@
+MDS Quiesce Protocol
+====================
+
+The MDS quiesce protocol is a mechanism for "quiescing" (quieting) a tree in a
+file system, stopping all write (and sometimes incidentally read) I/O.
+
+The purpose of this API is to prevent multiple clients from interleaving reads
+and writes across an eventually consistent snapshot barrier where out-of-band
+communication exists between clients. This communication can lead to clients
+wrongly believing they've reached a checkpoint that is mutually recoverable to
+via a snapshot.
+
+
+Mechanism
+---------
+
+The MDS quiesces I/O using a new ``quiesce_path`` internal request that obtains
+appropriate locks on the root of a tree and then launches a series of
+sub-requests for locking other inodes in the tree. The locks obtained will
+force clients to release caps and in-progress client/MDS requests to complete.
+
+The sub-requests launched are ``quiesce_inode`` internal requests. These will
+obtain "cap-related" locks which control capability state, including the
+``filelock``, ``authlock``, ``linklock``, and ``xattrlock``. Additionally, the
+new local lock ``quiescelock`` is acquired. More information on that lock in
+the next section.
+
+Locks that are not cap-related are skipped because they do not control typical
+and durable metadata state. Additionally, only Capabilities can give a client
+local control of a file's metadata or data.
+
+Once all locks have been acquired, the cap-related locks are released and the
+``quiescelock`` is relied on to prevent issuing Capabilities to clients for the
+cap-related locks. This is controlled primarily by ``CInode:get_caps_*``
+methods. Releasing these locks is necessary to allow other ranks with the
+replicated inode to quiesce without lock state transitions resulting in
+deadlock. For example, a client wanting ``Xx`` on an inode will trigger a
+``xattrlock`` in ``LOCK_SYNC`` state to transition to ``LOCK_SYNC_EXCL``.  That
+state would not allow another rank to acquire ``xattrlock`` for reading,
+thereby creating deadlock, subject to quiesce expiration. (Quiesce cannot
```

Comment:
... subject to quiesce timeout/expiration

Reply:
fixed